[18.0][ADD] fs_folder: An new addon to link a folder accessed through the fsspecs library to an odoo record.

lmignon opened the pull request:
If you need to link some specific models to a specific folder into an external filesystem and be able to manage the content of this folder from the model form view, the `fs_folder` addon is for you.
![image](https://github.com/user-attachments/assets/a1f2ae55-46e3-4a7c-a177-534527da741d)

With the contribution of @etobella

lmignon committed:
[ADD] fs_field: Provides fields for filesystem contents

lmignon committed:
[IMP] fs_field: Add api to be used by the FsFolder widget

lmignon committed:
[IMP] fs_field: Add documentation

lmignon committed:
[IMP] fs_field: Add cleanup method in case of rollback

lmignon committed:
[IMP] fs_field: Add a way to sanitize the path used to create a folder

lmignon committed:
[FIX] fs_field: makes tests consistent from one environment to another

lmignon committed:
[IMP] fs_field: Add missing api methods

lmignon committed:
[FIX] fs_field: doc

sbidoul committed:
[DOC] fs_field: fix doc typos

etobella committed:
[FIX] fs_content: Make preview work

etobella committed:
[ADD] fs_field_demo

etobella committed:
[IMP] fs_field: Add JS

etobella committed:
[IMP] fs_field: Pass logic to a service in order to use it in other places

etobella committed:
[IMP] fs_field: Allow to initialize from the widget directly

etobella committed:
[FIX] fs_field_demo: Hide button

etobella committed:
[IMP] fs_field: Improve views. Add headers

etobella committed:
[FIX] fs_field: Fix upload of files

etobella committed:
[IMP] fs_field: Remove double components. improve buttons

etobella committed:
[IMP] fs_field: Allow to rename folders & files

etobella committed:
[IMP] fs_field: Improve management of icons

etobella committed:
[IMP] fs_field: Improve table management

etobella committed:
[FIX] fs_field: Make button work

etobella committed:
[IMP] fs_field: Allow sort of files

etobella committed:
[IMP] fs_field: Allow to hide columns

etobella committed:
[FIX] fs_field: Use proper names

etobella committed:
[IMP] fs_field: Improve view

etobella committed:
[FIX] fs_field: use onWillRender

etobella committed:
[IMP] fs_field: Add names. Fix directory icon

lmignon committed:
[IMP] fs_field: Improves sorting

* Sort char and text using locale and in a case and accent incensitive way
* When soring on name, keep directory first into the list
* Allows to define a comparison method into the field definition used if specified into the sort method

lmignon committed:
[FIX] fs_field: Pass the right parameter to the adapter

etobella committed:
[IMP] fs_field: Use effect. initialize some states

etobella committed:
Merge pull request #1 from acsone/18.0-js-lmi

Some improvements

lmignon committed:
Move fs_folder and fs_folder_demo from incubator

lmignon committed:
[IMP] fs_storage_*: renamed from fs_field_*

etobella commented:
@lmignon on the rename, all commit history will be lost on next version. There are options to rename the history or you can just squash all commits together and add coauthors in order to simplify everything (second option is probably the easier one)

lmignon commented:
> @lmignon on the rename, all commit history will be lost on next version. There are options to rename the history or you can just squash all commits together and add coauthors in order to simplify everything (second option is probably the easier one)

No need to do it :smirk:; the path into the history is also renamed...
![image](https://github.com/user-attachments/assets/2c87a409-0148-4d1c-8a38-a71431310c86)

lmignon committed:
[IMP] fs_folder: Proper tests isolation

When creating a storage used to store fs folder content, disbale possible existing one created by others addons

lmignon committed:
[FIX] fs_folder: call to method initialize on FsFolderValue instance must return a new instance

lmignon committed:
[IMP] fs_folder: Get access to the fs_storage instance from FsFolder value

lmignon committed:
[IMP] fs_folder: Allows to notify the widged when the content is modified

This action is usefull when an action from within odoo update the content and the widget need to be refreshed

lmignon committed:
[IMP] fs_folder: Avoid error in tests

Uses the TransactionCase class in place of BaseCommon as base class for tests to avoid error when tests are run on an database where the account module is installed but not loaded into the registry. The account addon adds a required column on the res.partner model with a default value but if the module is not loaded into the registry when a res.partner record is created, the creation fails with an integrity error

lmignon committed:
[FIX] fs_folder: fix sanitize method

The hyphen is a valid character in the name of a directory or file.

lmignon committed:
[IMP] fs_folder: Improve breadcrumb display

youring commented:
Thank you for creating this very useful module! It seems very similar to Alfodoo project. Will version control be added in the future?

Tested on runboat, found two points: 
If create a folder for the first time, 
`ValueError: No default storage found for the content of the external filesystem fields for model res.partner and field fs_folder_field. Please set a default storage in the filesystem storage configuration.`
Go to **FS Storage**, set **Odoo Filesystem Backend** as **Default For Fs Contents**, now it's fixed.
Should this be automatically set by demo module?

It seems preview is handled by the standalone SimpleDialog method, image is not centered, with white frame, and it can't be closed with the ESC key.
<img width="1452" height="865" alt="preview_img_not_centered_white_bg" src="https://github.com/user-attachments/assets/62a0f032-062a-470f-90b0-727b73fac1d9" />
And it's unable to navigate multiple files in current folder using the left and right arrow keys like Odoo core's preview.
A dumb question: for consistency, can we leverage Odoo core's preview component for attachment preview to replace SimpleDialog here?

lmignon commented:
> Thank you for creating this very useful module! It seems very similar to Alfodoo project. 

Same company, same authors ... 

> Will version control be added in the future?

The “fs_folder” module allows you to manage content stored in file systems external to Odoo from within Odoo. In its basic version, it focuses solely on cross-functional features for all file systems supported by the Python fsspecs library, which we use as an abstraction layer for the various file systems. Document versioning is not one of these standard features and is therefore not offered. However, the module has been designed to be easily extensible. An initial specialisation offered concerns more specific support for Microsoft Drive ( and SharePoint) directories. In the latter case, the target file system offers document version management features. These could be exposed in this specific context, even if it is not currently a requirement expressed by our customers. But do we really need it into Odoo? Indeed, It's very easy for users to jump from Odoo to a specific file or directory in the Microsoft application that manages these items and benefit from these advanced features without having to complicate the implementation in Odoo.

> 
> Tested on runboat, found two points: If create a folder for the first time, `ValueError: No default storage found for the content of the external filesystem fields for model res.partner and field fs_folder_field. Please set a default storage in the filesystem storage configuration.` Go to **FS Storage**, set **Odoo Filesystem Backend** as **Default For Fs Contents**, now it's fixed. Should this be automatically set by demo module?

It could

> 
> It seems preview is handled by the standalone SimpleDialog method, image is not centered, with white frame, and it can't be closed with the ESC key. <img alt="preview_img_not_centered_white_bg" width="1452" height="865" src="https://private-user-images.githubusercontent.com/4132418/483071457-62a0f032-062a-470f-90b0-727b73fac1d9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYzODQxOTksIm5iZiI6MTc1NjM4Mzg5OSwicGF0aCI6Ii80MTMyNDE4LzQ4MzA3MTQ1Ny02MmEwZjAzMi0wNjJhLTQ3MGYtOTBiMC03MjdiNzNmYWMxZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjhUMTIyNDU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWI5NDUxYzVlMDQxYTkyNjU4MDFlZGE2ZTAwZDQyY2I3OGM1YzI3Y2ZkNGE2MWM1YzAyNzhiN2Y0YTRhNmFiMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.YQxIziqbjOhoYlMxrCrqkONQ0wTYDWbiR1pXN2iceF8"> And it's unable to navigate multiple files in current folder using the left and right arrow keys like Odoo core's preview. A dumb question: for consistency, can we leverage Odoo core's preview component for attachment preview to replace SimpleDialog here?

We have not found a simple way to use the native Odoo component. The preview could be improved, and all contributions are welcome. At this stage, we are not using the basic version of the `fs_folder` addon; we are using the `fs_folder_ms_drive` addon. Nevertheless, we have taken care to design this new document management approach to make it as modular as possible. As mentioned above, `fs_folder` is the basic component, and although it already offers many features, some of them, such as the preview, are offered in this first draft in a simple form because they are not used in our projects (in the specialization for ms_drive, we use the preview system provided by Microsoft).

lmignon commented:
/ocabot merge nobump
